File: SunoFlowApp/Sources/SunoFlow/Preferences.swift
import Carbon
import AppKit
import Foundation

extension Notification.Name {
    /// Posted when the dictation hotkey changes so the AppDelegate can re-register it.
    static let sunoHotkeyChanged = Notification.Name("suno.hotkeyChanged")
    /// Posted when the tone hotkey or its combination changes, so the
    /// AppDelegate can register or re-register it.
    static let sunoToneHotkeyChanged = Notification.Name("suno.toneHotkeyChanged")
    /// Posted when the Suno Answer hotkey or its combination changes, so the
    /// AppDelegate can register or re-register it.
    static let sunoAnswerHotkeyChanged = Notification.Name("suno.answerHotkeyChanged")
    /// Posted when the Suno Control hotkey or its combination changes, so the
    /// AppDelegate can register or re-register it.
    static let sunoControlHotkeyChanged = Notification.Name("suno.controlHotkeyChanged")
    /// Posted when the gateway refuses Suno Answer (not entitled). The
    /// AppDelegate opens the account sheet — the same surface a dictation 402
    /// lands on — carrying `message` and `code` in userInfo.
    static let sunoAnswerNotEntitled = Notification.Name("suno.answerNotEntitled")
}

/// Cross-process notifications (delivered via DistributedNotificationCenter).
enum AppNotifications {
    /// Sent by a second launch of the app to ask the running instance to open Settings.
    static let openSettings = Notification.Name("com.sunoapp.sunoflow.openSettings")
}

/// App-wide user settings, persisted in `UserDefaults` and observable from SwiftUI.
///
/// These back the values that used to be hardcoded across the app: the mic device,
/// the dictation hotkey, the auto-stop duration, and whether the AI cleanup pass
/// runs. The cleanup instruction and LLM model are owned by the hosted cleanup
/// gateway (see `TranscriptionClient.checkCleanupGateway`); the app only toggles
/// whether the cleanup pass runs and probes the gateway's reachability.
final class Preferences: ObservableObject {
    static let shared = Preferences()

    private let defaults = UserDefaults.standard

    private enum Key {
        static let micDeviceUID = "micDeviceUID"
        static let protectBluetoothAudio = "protectBluetoothAudio"
        static let hotkeyKeyCode = "hotkeyKeyCode"
        static let hotkeyModifiers = "hotkeyModifiers"
        static let maxRecordingSeconds = "maxRecordingSeconds"
        static let cleanupEnabled = "cleanupEnabled"
        static let tone = "tone"
        static let toneHotkeyEnabled = "toneHotkeyEnabled"
        static let toneHotkeyKeyCode = "toneHotkeyKeyCode"
        static let toneHotkeyModifiers = "toneHotkeyModifiers"
        static let answerHotkeyEnabled = "answerHotkeyEnabled"
        static let answerHotkeyKeyCode = "answerHotkeyKeyCode"
        static let answerHotkeyModifiers = "answerHotkeyModifiers"
        static let controlHotkeyEnabled = "controlHotkeyEnabled"
        static let controlHotkeyKeyCode = "controlHotkeyKeyCode"
        static let controlHotkeyModifiers = "controlHotkeyModifiers"
        static let offerCopyWhenUnfocused = "offerCopyWhenUnfocused"
        static let onboardingCompleted = "onboardingCompleted"
        static let appearance = "appearance"
    }

    /// Which appearance the app forces, independent of the system setting.
    enum AppearancePreference: String, CaseIterable, Identifiable {
        case system, light, dark
        var id: Self { self }
        var label: String {
            switch self {
            case .system: return "Match system"
            case .light:  return "Always light"
            case .dark:   return "Always dark"
            }
        }
    }

    /// Core Audio device UID to record from. Empty means "system default input".
    @Published var micDeviceUID: String {
        didSet { defaults.set(micDeviceUID, forKey: Key.micDeviceUID) }
    }

    /// Keep Bluetooth headsets out of the system's default *input* slot.
    ///
    /// macOS hands the input slot to earbuds the moment they connect, which
    /// switches them into call mode and drops their playback to 16 kHz mono in
    /// every app. On means SunoFlow moves capture back to the built-in mic —
    /// unless a call already has the headset mic open. See `BluetoothAudioGuard`.
    @Published var protectBluetoothAudio: Bool {
        didSet { defaults.set(protectBluetoothAudio, forKey: Key.protectBluetoothAudio) }
    }

    /// Carbon virtual key code for the dictation toggle hotkey.
    @Published var hotkeyKeyCode: UInt32 {
        didSet { persistHotkey() }
    }

    /// Carbon modifier mask (cmdKey / optionKey / …) for the dictation hotkey.
    @Published var hotkeyModifiers: UInt32 {
        didSet { persistHotkey() }
    }

    /// Auto-stop a runaway recording after this many seconds.
    @Published var maxRecordingSeconds: Int {
        didSet { defaults.set(maxRecordingSeconds, forKey: Key.maxRecordingSeconds) }
    }

    /// Run the local LLM cleanup pass. When off, the raw transcript is pasted.
    @Published var cleanupEnabled: Bool {
        didSet { defaults.set(cleanupEnabled, forKey: Key.cleanupEnabled) }
    }

    /// The writing voice the cleanup pass writes in. Sticky by design — it is
    /// cycled from a key while dictating, so resetting it every dictation would
    /// make the key feel like it had not registered. `Tone.faithful`, the
    /// default, sends nothing and leaves the user's own wording alone.
    ///
    /// Stored as the raw ID rather than an index: an index would silently mean
    /// a different voice the moment the list gains or loses an entry.
    @Published var tone: Tone {
        didSet { defaults.set(tone.rawValue, forKey: Key.tone) }
    }

    /// Cycle the tone with a second system-wide hotkey (⌥⇧Space by default).
    ///
    /// Off by default: the dictation hotkey already does something with every
    /// press, and a second global shortcut should be an opt-in. On, it is an
    /// ordinary Carbon hotkey — reliable system-wide, unlike the old Fn tap.
    @Published var toneHotkeyEnabled: Bool {
        didSet {
            defaults.set(toneHotkeyEnabled, forKey: Key.toneHotkeyEnabled)
            NotificationCenter.default.post(name: .sunoToneHotkeyChanged, object: nil)
        }
    }

    /// Carbon virtual key code for the tone-cycle hotkey.
    @Published var toneHotkeyKeyCode: UInt32 {
        didSet { persistToneHotkey() }
    }

    /// Carbon modifier mask (cmdKey / optionKey / …) for the tone-cycle hotkey.
    @Published var toneHotkeyModifiers: UInt32 {
        didSet { persistToneHotkey() }
    }

    /// Show the Suno Answer popup (⌃⌥Space by default). Off by default because
    /// turning it on is the consent step (E1/E2): the toggle's description says
    /// exactly what leaves the Mac — a screen snapshot attached to each
    /// question — and nothing is captured while the feature is off.
    @Published var answerHotkeyEnabled: Bool {
        didSet {
            defaults.set(answerHotkeyEnabled, forKey: Key.answerHotkeyEnabled)
            NotificationCenter.default.post(name: .sunoAnswerHotkeyChanged, object: nil)
        }
    }

    /// Carbon virtual key code for the Suno Answer hotkey.
    @Published var answerHotkeyKeyCode: UInt32 {
        didSet { persistAnswerHotkey() }
    }

    /// Carbon modifier mask (cmdKey / optionKey / …) for the Suno Answer hotkey.
    @Published var answerHotkeyModifiers: UInt32 {
        didSet { persistAnswerHotkey() }
    }

    /// Run the Suno Control agent loop (⌃⌥⇧Space by default). Off by default
    /// because turning it on is the consent step: the feature looks at the
    /// screen and moves the cursor, so nothing is captured while it is off.
    @Published var controlHotkeyEnabled: Bool {
        didSet {
            defaults.set(controlHotkeyEnabled, forKey: Key.controlHotkeyEnabled)
            NotificationCenter.default.post(name: .sunoControlHotkeyChanged, object: nil)
        }
    }

    /// Carbon virtual key code for the Suno Control hotkey.
    @Published var controlHotkeyKeyCode: UInt32 {
        didSet { persistControlHotkey() }
    }

    /// Carbon modifier mask (cmdKey / optionKey / …) for the Suno Control hotkey.
    @Published var controlHotkeyModifiers: UInt32 {
        didSet { persistControlHotkey() }
    }

    /// When a finished dictation has nowhere to paste — no text field focused,
    /// or no Accessibility permission to press Cmd+V — show the transcript at the
    /// bottom of the screen with a button that copies it, instead of typing into
    /// nothing. Off returns to pasting blindly and hoping something catches it.
    @Published var offerCopyWhenUnfocused: Bool {
        didSet { defaults.set(offerCopyWhenUnfocused, forKey: Key.offerCopyWhenUnfocused) }
    }

    /// False until first-run setup has been completed (or explicitly skipped).
    /// Stored rather than inferred from "is everything configured", so someone
    /// who deletes their model later gets the model page, not the whole wizard.
    @Published var onboardingCompleted: Bool {
        didSet { defaults.set(onboardingCompleted, forKey: Key.onboardingCompleted) }
    }

    /// The app's forced appearance (independent of the system's). `nil` follows
    /// the system; the other two pin `NSApp.appearance` accordingly.
    @Published var appearance: AppearancePreference {
        didSet {
            defaults.set(appearance.rawValue, forKey: Key.appearance)
            applyAppearance()
        }
    }

    /// Applies the chosen appearance to the whole app. The panels are AppKit
    /// views with dynamic colors plus a `viewDidChangeEffectiveAppearance`
    /// re-paint, and the settings window is SwiftUI — both re-resolve
    /// automatically when `NSApp.appearance` changes.
    private func applyAppearance() {
        switch appearance {
        case .system: NSApp.appearance = nil
        case .light:  NSApp.appearance = NSAppearance(named: .aqua)
        case .dark:   NSApp.appearance = NSAppearance(named: .darkAqua)
        }
    }

    private init() {
        defaults.register(defaults: [
            Key.micDeviceUID: "",
            Key.protectBluetoothAudio: true,
            Key.hotkeyKeyCode: Int(DefaultHotkey.keyCode),
            Key.hotkeyModifiers: Int(DefaultHotkey.modifiers),
            Key.maxRecordingSeconds: 300,
            Key.cleanupEnabled: true,
            Key.tone: Tone.faithful.rawValue,
            Key.toneHotkeyEnabled: false,
            Key.toneHotkeyKeyCode: Int(DefaultToneHotkey.keyCode),
            Key.toneHotkeyModifiers: Int(DefaultToneHotkey.modifiers),
            Key.answerHotkeyEnabled: false,
            Key.answerHotkeyKeyCode: Int(DefaultAnswerHotkey.keyCode),
            Key.answerHotkeyModifiers: Int(DefaultAnswerHotkey.modifiers),
            Key.controlHotkeyEnabled: false,
            Key.controlHotkeyKeyCode: Int(DefaultControlHotkey.keyCode),
            Key.controlHotkeyModifiers: Int(DefaultControlHotkey.modifiers),
            Key.offerCopyWhenUnfocused: true,
            Key.onboardingCompleted: false,
            Key.appearance: AppearancePreference.system.rawValue,
        ])
        // didSet does not fire for assignments inside init, so these load the
        // stored values without redundantly writing them back.
        micDeviceUID = defaults.string(forKey: Key.micDeviceUID) ?? ""
        protectBluetoothAudio = defaults.bool(forKey: Key.protectBluetoothAudio)
        hotkeyKeyCode = UInt32(defaults.integer(forKey: Key.hotkeyKeyCode))
        hotkeyModifiers = UInt32(defaults.integer(forKey: Key.hotkeyModifiers))
        maxRecordingSeconds = defaults.integer(forKey: Key.maxRecordingSeconds)
        cleanupEnabled = defaults.bool(forKey: Key.cleanupEnabled)
        tone = Tone.from(defaults.string(forKey: Key.tone))
        toneHotkeyEnabled = defaults.bool(forKey: Key.toneHotkeyEnabled)
        toneHotkeyKeyCode = UInt32(defaults.integer(forKey: Key.toneHotkeyKeyCode))
        toneHotkeyModifiers = UInt32(defaults.integer(forKey: Key.toneHotkeyModifiers))
        answerHotkeyEnabled = defaults.bool(forKey: Key.answerHotkeyEnabled)
        answerHotkeyKeyCode = UInt32(defaults.integer(forKey: Key.answerHotkeyKeyCode))
        answerHotkeyModifiers = UInt32(defaults.integer(forKey: Key.answerHotkeyModifiers))
        controlHotkeyEnabled = defaults.bool(forKey: Key.controlHotkeyEnabled)
        controlHotkeyKeyCode = UInt32(defaults.integer(forKey: Key.controlHotkeyKeyCode))
        controlHotkeyModifiers = UInt32(defaults.integer(forKey: Key.controlHotkeyModifiers))
        offerCopyWhenUnfocused = defaults.bool(forKey: Key.offerCopyWhenUnfocused)
        onboardingCompleted = defaults.bool(forKey: Key.onboardingCompleted)
        appearance = AppearancePreference(rawValue: defaults.string(forKey: Key.appearance) ?? "")
            ?? .system
        // didSet never fires during init, so apply the persisted choice now.
        applyAppearance()
    }

    private func persistHotkey() {
        defaults.set(Int(hotkeyKeyCode), forKey: Key.hotkeyKeyCode)
        defaults.set(Int(hotkeyModifiers), forKey: Key.hotkeyModifiers)
        NotificationCenter.default.post(name: .sunoHotkeyChanged, object: nil)
    }

    private func persistToneHotkey() {
        defaults.set(Int(toneHotkeyKeyCode), forKey: Key.toneHotkeyKeyCode)
        defaults.set(Int(toneHotkeyModifiers), forKey: Key.toneHotkeyModifiers)
        NotificationCenter.default.post(name: .sunoToneHotkeyChanged, object: nil)
    }

    private func persistAnswerHotkey() {
        defaults.set(Int(answerHotkeyKeyCode), forKey: Key.answerHotkeyKeyCode)
        defaults.set(Int(answerHotkeyModifiers), forKey: Key.answerHotkeyModifiers)
        NotificationCenter.default.post(name: .sunoAnswerHotkeyChanged, object: nil)
    }

    private func persistControlHotkey() {
        defaults.set(Int(controlHotkeyKeyCode), forKey: Key.controlHotkeyKeyCode)
        defaults.set(Int(controlHotkeyModifiers), forKey: Key.controlHotkeyModifiers)
        NotificationCenter.default.post(name: .sunoControlHotkeyChanged, object: nil)
    }

    /// Restore the built-in ⌥Space shortcut.
    func resetHotkeyToDefault() {
        hotkeyKeyCode = DefaultHotkey.keyCode
        hotkeyModifiers = DefaultHotkey.modifiers
    }

    /// Restore the built-in ⌥⇧Space tone shortcut.
    func resetToneHotkeyToDefault() {
        toneHotkeyKeyCode = DefaultToneHotkey.keyCode
        toneHotkeyModifiers = DefaultToneHotkey.modifiers
    }

    /// Restore the built-in ⌃⌥Space Suno Answer shortcut.
    func resetAnswerHotkeyToDefault() {
        answerHotkeyKeyCode = DefaultAnswerHotkey.keyCode
        answerHotkeyModifiers = DefaultAnswerHotkey.modifiers
    }

    /// Restore the built-in ⌃⌥⇧Space Suno Control shortcut.
    func resetControlHotkeyToDefault() {
        controlHotkeyKeyCode = DefaultControlHotkey.keyCode
        controlHotkeyModifiers = DefaultControlHotkey.modifiers
    }
}

extension Preferences {
    /// Advance to the next voice and hand back what it landed on, so the caller
    /// can announce it. One place for the cycle, whether it was driven from the
    /// key or from the menu.
    @discardableResult
    func cycleTone() -> Tone {
        tone = tone.next
        return tone
    }
}

File: apps/windows/SunoFlow/Preferences.cs
using System;
using System.ComponentModel;
using System.Text.Json;

namespace SunoFlow;

/// <summary>
/// App-wide user settings persisted as JSON in
/// <c>%LOCALAPPDATA%/SunoFlow/preferences.json</c>. Mirrors the macOS
/// <c>Preferences.swift</c> (which uses <c>UserDefaults</c>) — same fields,
/// same defaults, same "cleanup is server-hosted" stance. The hotkey here is a
/// (virtual-key code, modifier flags) pair, matching how Windows global
/// hotkeys work (see <see cref="HotkeyManager"/>).
/// </summary>
internal sealed class Preferences : INotifyPropertyChanged
{
    public static Preferences Instance { get; } = Load();

    private const int DefaultHotkeyCode = 0x20; // VK_SPACE
    private const int DefaultHotkeyMods = ModAlt;

    // Tone cycle: Alt+Shift+Space, the Windows analogue of the Mac's ⌥⇧Space.
    private const int DefaultToneHotkeyCode = 0x20; // VK_SPACE
    private const int DefaultToneHotkeyMods = ModAlt | ModShift;
    // Used when the default collides with a customised dictation shortcut:
    // Ctrl+Shift+Space, matching the Mac's ⌘⇧Space fallback.
    public const int FallbackToneHotkeyCode = 0x20;
    public const int FallbackToneHotkeyMods = ModControl | ModShift;

    // Win32 modifier flags for RegisterHotKey.
    public const int ModNone = 0x0000;
    public const int ModAlt = 0x0001;
    public const int ModControl = 0x0002;
    public const int ModShift = 0x0004;
    public const int ModWin = 0x0008;

    public string MicDeviceId { get; set; } = "";
    public int MaxRecordingSeconds { get; set; } = 60;
    public bool CleanupEnabled { get; set; } = true;
    public bool ScreenContextEnabled { get; set; } = false;

    /// <summary>When a dictation finishes and no text field has focus, offer the
    /// transcript instead of pasting it into nothing. On by default — the
    /// alternative is losing what the user just said.</summary>
    public bool OfferCopyWhenUnfocused { get; set; } = true;

    /// <summary>Steer capture away from a Bluetooth headset's call-quality mode.
    /// On by default, and only ever acts when the selected input really is a
    /// Bluetooth device.</summary>
    public bool ProtectBluetoothAudio { get; set; } = true;

    /// <summary>The writing voice, stored as the gateway's raw id rather than an
    /// index: an index would silently mean a different voice the moment the list
    /// gains or loses an entry. Serialized as a string so the JSON stays
    /// readable and matches what goes over the wire.</summary>
    public string ToneId { get; set; } = "";

    /// <summary>The chosen voice. Not serialized — <see cref="ToneId"/> is.</summary>
    [System.Text.Json.Serialization.JsonIgnore]
    public Tone Tone
    {
        get => Tone.From(ToneId);
        set { ToneId = value.Id; Save(); OnPropertyChanged(nameof(Tone)); }
    }

    /// <summary>Advance to the next voice and hand back what it landed on, so the
    /// caller can announce it. One place for the cycle, whether it was driven
    /// from the key or from the tray menu.</summary>
    public Tone CycleTone()
    {
        Tone = Tone.Next();
        return Tone;
    }

    /// <summary>False until first-run setup has been completed (or explicitly
    /// skipped). Stored rather than inferred from "is everything configured",
    /// so someone who deletes their model later gets the model page, not the
    /// whole wizard again.</summary>
    public bool OnboardingCompleted { get; set; } = false;

    // The hotkey pair raises PropertyChanged so the TrayApp can re-register the
    // system-wide shortcut the moment the user picks a new combo in Settings.
    private int _hotkeyCode = DefaultHotkeyCode;
    private int _hotkeyModifiers = DefaultHotkeyMods;
    public int HotkeyCode
    {
        get => _hotkeyCode;
        set { _hotkeyCode = value; Save(); OnPropertyChanged(nameof(HotkeyCode)); }
    }
    public int HotkeyModifiers
    {
        get => _hotkeyModifiers;
        set { _hotkeyModifiers = value; Save(); OnPropertyChanged(nameof(HotkeyModifiers)); }
    }

    // The tone-cycle hotkey. Off by default: the dictation hotkey already does
    // something with every press, and a second system-wide shortcut should be
    // something the user opts into. Raises PropertyChanged for the same reason
    // as the pair above — TrayApp re-registers on the change.
    private bool _toneHotkeyEnabled;
    private int _toneHotkeyCode = DefaultToneHotkeyCode;
    private int _toneHotkeyModifiers = DefaultToneHotkeyMods;

    public bool ToneHotkeyEnabled
    {
        get => _toneHotkeyEnabled;
        set { _toneHotkeyEnabled = value; Save(); OnPropertyChanged(nameof(ToneHotkeyEnabled)); }
    }
    public int ToneHotkeyCode
    {
        get => _toneHotkeyCode;
        set { _toneHotkeyCode = value; Save(); OnPropertyChanged(nameof(ToneHotkeyCode)); }
    }
    public int ToneHotkeyModifiers
    {
        get => _toneHotkeyModifiers;
        set { _toneHotkeyModifiers = value; Save(); OnPropertyChanged(nameof(ToneHotkeyModifiers)); }
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    private void OnPropertyChanged(string name) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));

    /// <summary>Restore the built-in Alt+Space shortcut.</summary>
    public void ResetHotkeyToDefault()
    {
        HotkeyCode = DefaultHotkeyCode;
        HotkeyModifiers = DefaultHotkeyMods;
        // The setters already save + raise PropertyChanged for both fields.
    }

    /// <summary>Restore the built-in Alt+Shift+Space tone shortcut.</summary>
    public void ResetToneHotkeyToDefault()
    {
        ToneHotkeyCode = DefaultToneHotkeyCode;
        ToneHotkeyModifiers = DefaultToneHotkeyMods;
    }

    /// <summary>Set while <see cref="Load"/> is populating a fresh instance. The
    /// hotkey setters persist on write, and during a load that would serialize the
    /// half-populated object straight back over the file being read — whatever had
    /// not been assigned yet would silently revert to its default.</summary>
    private static bool _loading;

    public void Save()
    {
        if (_loading) return;
        try
        {
            var dir = System.IO.Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "SunoFlow");
            System.IO.Directory.CreateDirectory(dir);
            var path = System.IO.Path.Combine(dir, "preferences.json");
            var json = JsonSerializer.Serialize(this, JsonOpts);

            // Write beside the real file and swap it in, so a crash or a full disk
            // mid-write leaves the previous settings intact rather than a truncated
            // file the next launch cannot parse.
            var temp = path + ".tmp";
            System.IO.File.WriteAllText(temp, json);
            System.IO.File.Move(temp, path, overwrite: true);
        }
        catch (Exception ex)
        {
            AppLog.Log($"Failed to save preferences: {ex.Message}");
        }
    }

    private static Preferences Load()
    {
        try
        {
            var path = System.IO.Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "SunoFlow", "preferences.json");
            if (System.IO.File.Exists(path))
            {
                var json = System.IO.File.ReadAllText(path);
                _loading = true;
                try
                {
                    var p = JsonSerializer.Deserialize<Preferences>(json, JsonOpts);
                    if (p != null) return p;
                }
                finally { _loading = false; }
            }
        }
        catch (Exception ex)
        {
            AppLog.Log($"Failed to load preferences, using defaults: {ex.Message}");
        }
        return new Preferences();
    }

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        // We write camelCase but the C# props are PascalCase, so the loader must
        // match case-insensitively or saved settings would be silently dropped.
        PropertyNameCaseInsensitive = true,
    };
}